File: Component/BusiLib/src/main/java-person/com/component/person/producation/view/ProducationWallView.java
package com.component.person.producation.view;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.SpannableStringBuilder;
import android.view.Gravity;
import android.view.View;
import android.widget.RelativeLayout;

import com.alibaba.fastjson.JSON;
import com.common.base.BaseFragment;
import com.common.core.myinfo.MyUserInfoManager;
import com.common.core.userinfo.UserInfoServerApi;
import com.common.core.userinfo.model.UserInfoModel;
import com.common.player.IPlayer;
import com.common.player.MyMediaPlayer;
import com.common.player.VideoPlayerAdapter;
import com.common.rxretrofit.ApiManager;
import com.common.rxretrofit.ApiMethods;
import com.common.rxretrofit.ApiObserver;
import com.common.rxretrofit.ApiResult;
import com.common.utils.SpanUtils;
import com.common.view.DebounceViewClickListener;
import com.component.busilib.R;
import com.component.person.producation.adapter.ProducationAdapter;
import com.component.person.producation.model.ProducationModel;
import com.component.person.view.RequestCallBack;
import com.dialog.view.TipsDialogView;
import com.orhanobut.dialogplus.DialogPlus;
import com.orhanobut.dialogplus.ViewHolder;
import com.component.dialog.ShareWorksDialog;

import java.util.HashMap;
import java.util.List;

import okhttp3.MediaType;
import okhttp3.RequestBody;

/**
 * 作品墙view
 */
public class ProducationWallView extends RelativeLayout {

    public final String TAG = "ProducationWallView";

    BaseFragment mFragment;
    RequestCallBack mCallBack;
    UserInfoServerApi mUserInfoServerApi;
    UserInfoModel mUserInfoModel;

    RecyclerView mProducationView;
    ProducationAdapter mAdapter;

    IPlayer mIPlayer;

    int DEFAUAT_CNT = 20;       // 默认拉取一页的数量
    int offset;  // 拉照片偏移量
    long mLastUpdateInfo = 0;    //上次更新成功时间

    DialogPlus mConfirmDialog;
    ShareWorksDialog mShareWorksDialog;

    public UserInfoModel getUserInfoModel() {
        return mUserInfoModel;
    }

    public void setUserInfoModel(UserInfoModel userInfoModel) {
        mUserInfoModel = userInfoModel;
    }

    public ProducationWallView(BaseFragment fragment, UserInfoModel userInfoModel, RequestCallBack requestCallBack) {
        super(fragment.getContext());
        this.mFragment = fragment;
        this.mUserInfoModel = userInfoModel;
        this.mCallBack = requestCallBack;
        mUserInfoServerApi = ApiManager.getInstance().createService(UserInfoServerApi.class);
        init();
    }

    private void init() {
        inflate(getContext(), R.layout.producation_wall_view_layout, this);

        mProducationView = findViewById(R.id.producation_view);
        boolean isSelf = false;
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);
        mProducationView.setLayoutManager(linearLayoutManager);
        if (mUserInfoModel != null && mUserInfoModel.getUserId() == MyUserInfoManager.getInstance().getUid()) {
            isSelf = true;
        }
        mAdapter = new ProducationAdapter(new ProducationAdapter.Listener() {
            @Override
            public void onClickDele(int position, ProducationModel model) {
                if (model.getWorksID() == mAdapter.getPlayingWorksIdPosition()) {
                    // 先停止播放
                    stopPlay();
                }
                // TODO: 2019/5/22 弹出删除确认框
                showConfirmDialog(model);
            }

            @Override
            public void onClickShare(int position, ProducationModel model) {
                // TODO: 2019/5/22 弹出分享框 需不需要先停止音乐
                if (model.getWorksID() == mAdapter.getPlayingWorksIdPosition()) {
                    // 先停止播放
                    stopPlay();
                }
                showShareDialog(model);
            }

            @Override
            public void onClickPlayBtn(View view, boolean play, int position, ProducationModel model) {
                if (play) {
                    if (mIPlayer == null) {
                        mIPlayer = new MyMediaPlayer();
                        mIPlayer.setDecreaseVolumeEnd(true);
                        // 播放完毕
                        mIPlayer.setCallback(new VideoPlayerAdapter.PlayerCallbackAdapter() {
                            @Override
                            public void onCompletion() {
                                super.onCompletion();
                                mAdapter.setPlayPosition(-1);
                            }
                        });
                    }
                    mIPlayer.reset();
                    mIPlayer.startPlay(model.getWorksURL());
                    playProducation(model, position);
                    // 开始播放当前postion，
                    // 清楚上一个
                    mAdapter.setPlayPosition(model.getWorksID());
                } else {
                    if (mIPlayer != null) {
                        //mIPlayer.setCallback(null);
                        mIPlayer.pause();
                    }
                    // 不用刷新，优化下，防止闪动， icon 在 click 事件内部已经设置过了
                    mAdapter.setPlayPosition(-1);
                }
            }
        }, isSelf, linearLayoutManager);
        mProducationView.setAdapter(mAdapter);
    }

    private void showConfirmDialog(final ProducationModel model) {
        SpannableStringBuilder stringBuilder = new SpanUtils()
                .append("确定删除该作品吗？")
                .create();
        TipsDialogView tipsDialogView = new TipsDialogView.Builder(getContext())
                .setMessageTip(stringBuilder)
                .setConfirmTip("确认")
                .setCancelTip("我再想想")
                .setConfirmBtnClickListener(new DebounceViewClickListener() {
                    @Override
                    public void clickValid(View v) {
                        if (mConfirmDialog != null) {
                            mConfirmDialog.dismiss();
                        }
                        deleteProducation(model);
                    }
                })
                .setCancelBtnClickListener(new DebounceViewClickListener() {
                    @Override
                    public void clickValid(View v) {
                        if (mConfirmDialog != null) {
                            mConfirmDialog.dismiss();
                        }
                    }
                })
                .build();

        mConfirmDialog = DialogPlus.newDialog(getContext())
                .setContentHolder(new ViewHolder(tipsDialogView))
                .setGravity(Gravity.BOTTOM)
                .setContentBackgroundResource(R.color.transparent)
                .setOverlayBackgroundResource(R.color.black_trans_80)
                .setExpanded(false)
                .create();
        mConfirmDialog.show();

    }

    private void showShareDialog(final ProducationModel model) {
        if (mShareWorksDialog != null) {
            mShareWorksDialog.dismiss(false);
        }
        mShareWorksDialog = new ShareWorksDialog(mFragment, model.getName(), false);
        mShareWorksDialog.setData(model.getUserID(), model.getNickName(), model.getCover(), model.getName(), model.getWorksURL(), model.getWorksID());
        mShareWorksDialog.show();
    }

    public void stopPlay() {
        mAdapter.setPlayPosition(-1);
        if (mIPlayer != null) {
            mIPlayer.stop();
        }
    }

    public void getProducations(boolean isFlag) {
        long now = System.currentTimeMillis();
        if (!isFlag) {
            // 10分钟更新一次吧
            if ((now - mLastUpdateInfo) < 10 * 60 * 1000) {
                return;
            }
        }

        getProducations(0);
    }

    public void getMoreProducations() {
        getProducations(offset);
    }

    private void getProducations(final int offset) {
        ApiMethods.subscribe(mUserInfoServerApi.getWorks(mUserInfoModel.getUserId(), offset, DEFAUAT_CNT), new ApiObserver<ApiResult>() {
            @Override
            public void process(ApiResult result) {
                if (result.getErrno() == 0) {
                    int totalCnt = result.getData().getIntValue("totalCnt");
                    int newOffset = result.getData().getIntValue("offset");
                    List<ProducationModel> list = JSON.parseArray(result.getData().getString("works"), ProducationModel.class);
                    if (offset == 0) {
                        addProducation(list, newOffset, totalCnt, true);
                    } else {
                        addProducation(list, newOffset, totalCnt, false);
                    }
                } else {
                    loadProducationsFailed();
                }
            }

            @Override
            public void onNetworkError(ErrorType errorType) {
                super.onNetworkError(errorType);
                loadProducationsFailed();
            }
        }, mFragment);
    }


    private void deleteProducation(final ProducationModel model) {
        HashMap<String, Object> map = new HashMap<>();
        map.put("worksID", model.getWorksID());
        RequestBody body = RequestBody.create(MediaType.parse(ApiManager.APPLICATION_JSON), JSON.toJSONString(map));
        ApiMethods.subscribe(mUserInfoServerApi.deleWorks(body), new ApiObserver<ApiResult>() {
            @Override
            public void process(ApiResult result) {
                if (result.getErrno() == 0) {
                    mAdapter.delete(model);
                    mAdapter.notifyDataSetChanged();
                }
            }
        }, mFragment);
    }

    public void playProducation(final ProducationModel model, final int position) {
        HashMap<String, Object> map = new HashMap<>();
        map.put("toUserID", mUserInfoModel.getUserId());
        map.put("worksID", model.getWorksID());
        RequestBody body = RequestBody.create(MediaType.parse(ApiManager.APPLICATION_JSON), JSON.toJSONString(map));
        ApiMethods.subscribe(mUserInfoServerApi.playWorks(body), new ApiObserver<ApiResult>() {
            @Override
            public void process(ApiResult result) {
                if (result.getErrno() == 0) {
                    // TODO: 2019/5/22 播放次数客户端自己加一
                    model.setPlayCnt(model.getPlayCnt() + 1);
                    mAdapter.notifyDataSetChanged();
                }
            }
        }, mFragment, new ApiMethods.RequestControl("playWorks", ApiMethods.ControlType.CancelThis));
    }

    private void addProducation(List<ProducationModel> list, int newOffset, int totalCnt, boolean isClear) {
        offset = newOffset;
        mLastUpdateInfo = System.currentTimeMillis();

        if (mCallBack != null) {
            mCallBack.onRequestSucess();
        }
        if (isClear) {
            mAdapter.getDataList().clear();
        }

        if (list != null && list.size() > 0) {
            mAdapter.getDataList().addAll(list);
            mAdapter.notifyDataSetChanged();
        } else {
            if (mAdapter.getDataList() != null && mAdapter.getDataList().size() > 0) {
                // 没有更多了
            } else {
                // 没有数据
            }
        }
    }

    private void loadProducationsFailed() {
        if (mCallBack != null) {
            mCallBack.onRequestSucess();
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if (mIPlayer != null) {
            mIPlayer.setCallback(null);
            mIPlayer.stop();
            mIPlayer.release();
        }
        if (mShareWorksDialog != null) {
            mShareWorksDialog.dismiss(false);
        }
        if (mConfirmDialog != null) {
            mConfirmDialog.dismiss(false);
        }
    }
}

File: Component/BusiLib/src/main/java-person/com/component/person/producation/view/ProducationWallView.kt
package com.component.person.producation.view

import android.media.MediaPlayer
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.Gravity
import android.view.View
import android.widget.RelativeLayout
import com.alibaba.fastjson.JSON
import com.common.base.BaseFragment
import com.common.core.myinfo.MyUserInfoManager
import com.common.core.userinfo.UserInfoServerApi
import com.common.core.userinfo.model.UserInfoModel
import com.common.player.PlayerCallbackAdapter
import com.common.player.SinglePlayer
import com.common.rxretrofit.*
import com.common.utils.SpanUtils
import com.common.view.DebounceViewClickListener
import com.component.busilib.R
import com.component.dialog.ShareWorksDialog
import com.component.person.producation.adapter.ProducationAdapter
import com.component.person.producation.model.ProducationModel
import com.component.person.view.RequestCallBack
import com.dialog.view.TipsDialogView
import com.orhanobut.dialogplus.DialogPlus
import com.orhanobut.dialogplus.ViewHolder
import okhttp3.MediaType
import okhttp3.RequestBody
import java.util.*

/**
 * 作品墙view
 */
class ProducationWallView(internal var mFragment: BaseFragment, var userInfoModel: UserInfoModel, internal var mCallBack: RequestCallBack?) : RelativeLayout(mFragment.context) {

    val TAG = "ProducationWallView"
    val playerTag = TAG + hashCode()
    internal val mUserInfoServerApi: UserInfoServerApi = ApiManager.getInstance().createService(UserInfoServerApi::class.java)

    private val mProducationView: RecyclerView
    internal val mAdapter: ProducationAdapter

    private var DEFAUAT_CNT = 20       // 默认拉取一页的数量
    internal var offset: Int = 0  // 拉照片偏移量
    private var mLastUpdateInfo: Long = 0    //上次更新成功时间
    var hasMore = true

    internal var mConfirmDialog: DialogPlus? = null
    private var mShareWorksDialog: ShareWorksDialog? = null
    val playCallback: PlayerCallbackAdapter

    init {

        View.inflate(context, R.layout.producation_wall_view_layout, this)
        mProducationView = findViewById(R.id.producation_view)
        val linearLayoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        mProducationView.layoutManager = linearLayoutManager

        var isSelf = false
        if (userInfoModel.userId.toLong() == MyUserInfoManager.getInstance().uid) {
            isSelf = true
        }

        mAdapter = ProducationAdapter(isSelf)
        mAdapter.mOnClickDeleListener = { _, model ->
            if (model?.worksID == mAdapter.playingWorksIdPosition) {
                // 先停止播放
                stopPlay()
            }
            model?.let { showConfirmDialog(it) }
        }
        mAdapter.mOnClickShareListener = { _, model ->
            // TODO: 2019/5/22 弹出分享框 需不需要先停止音乐
            if (model?.worksID == mAdapter.playingWorksIdPosition) {
                // 先停止播放
                stopPlay()
            }
            model?.let { showShareDialog(it) }
        }

        playCallback = object : PlayerCallbackAdapter() {
            override fun onCompletion() {
                super.onCompletion()
                mAdapter.setPlayPosition(-1)
            }

            override fun openTimeFlyMonitor(): Boolean {
                return true
            }

            override fun onBufferingUpdate(mp: MediaPlayer?, percent: Int) {

            }

            override fun onTimeFlyMonitor(pos: Long, duration: Long) {

            }
        }
        SinglePlayer.addCallback(playerTag, playCallback)

        mAdapter.mOnClickPlayListener = { view, play, position, model ->
            if (play) {
                if (model != null) {
                    SinglePlayer.startPlay(playerTag, model?.worksURL ?: "")
                    playProducation(model, position)
                    // 开始播放当前postion，
                    // 清楚上一个
                    mAdapter.setPlayPosition(model.worksID)
                }
            } else {
                SinglePlayer.pause(playerTag)
                // 不用刷新，优化下，防止闪动， icon 在 click 事件内部已经设置过了
                mAdapter.setPlayPosition(-1)
            }
        }
        mProducationView.adapter = mAdapter
    }

    private fun showConfirmDialog(model: ProducationModel) {
        val stringBuilder = SpanUtils()
                .append("确定删除该作品吗？")
                .create()
        val tipsDialogView = TipsDialogView.Builder(context)
                .setMessageTip(stringBuilder)
                .setConfirmTip("确认")
                .setCancelTip("我再想想")
                .setConfirmBtnClickListener(object : DebounceViewClickListener() {
                    override fun clickValid(v: View) {
                        if (mConfirmDialog != null) {
                            mConfirmDialog!!.dismiss()
                        }
                        deleteProducation(model)
                    }
                })
                .setCancelBtnClickListener(object : DebounceViewClickListener() {
                    override fun clickValid(v: View) {
                        if (mConfirmDialog != null) {
                            mConfirmDialog!!.dismiss()
                        }
                    }
                })
                .build()

        mConfirmDialog = DialogPlus.newDialog(context)
                .setContentHolder(ViewHolder(tipsDialogView))
                .setGravity(Gravity.BOTTOM)
                .setContentBackgroundResource(R.color.transparent)
                .setOverlayBackgroundResource(R.color.black_trans_80)
                .setExpanded(false)
                .create()
        mConfirmDialog!!.show()

    }

    private fun showShareDialog(model: ProducationModel) {
        mShareWorksDialog?.dismiss(false)
        mShareWorksDialog = ShareWorksDialog(mFragment, model.name, false)
        mShareWorksDialog?.setData(model.userID, model.nickName, model.cover, model.name, model.worksURL, model.worksID)
        mShareWorksDialog?.show()
    }

    fun stopPlay() {
        mAdapter.setPlayPosition(-1)
        SinglePlayer.stop(playerTag)
    }

    fun getProducations(isFlag: Boolean) {
        val now = System.currentTimeMillis()
        if (!isFlag && mAdapter.dataList.isNotEmpty()) {
            // 10分钟更新一次吧
            if (now - mLastUpdateInfo < 10 * 60 * 1000) {
                return
            }
        }

        getProducations(0)
    }

    fun getMoreProducations() {
        getProducations(offset)
    }

    private fun getProducations(offset: Int) {
        ApiMethods.subscribe(mUserInfoServerApi.getWorks(userInfoModel.userId, offset, DEFAUAT_CNT), object : ApiObserver<ApiResult>() {
            override fun process(result: ApiResult) {
                if (result.errno == 0) {
                    val totalCnt = result.data!!.getIntValue("totalCnt")
                    val newOffset = result.data!!.getIntValue("offset")
                    val list = JSON.parseArray(result.data!!.getString("works"), ProducationModel::class.java)
                    if (offset == 0) {
                        addProducation(list, newOffset, totalCnt, true)
                    } else {
                        addProducation(list, newOffset, totalCnt, false)
                    }
                } else {
                    loadProducationsFailed()
                }
            }

            override fun onNetworkError(errorType: ApiObserver.ErrorType) {
                super.onNetworkError(errorType)
                loadProducationsFailed()
            }
        }, mFragment)
    }


    private fun deleteProducation(model: ProducationModel) {
        val map = HashMap<String, Any>()
        map["worksID"] = model.worksID
        val body = RequestBody.create(MediaType.parse(ApiManager.APPLICATION_JSON), JSON.toJSONString(map))
        ApiMethods.subscribe(mUserInfoServerApi.deleWorks(body), object : ApiObserver<ApiResult>() {
            override fun process(result: ApiResult) {
                if (result.errno == 0) {
                    mAdapter.delete(model)
                    mAdapter.notifyDataSetChanged()
                }
            }
        }, mFragment)
    }

    fun playProducation(model: ProducationModel, position: Int) {
        val map = HashMap<String, Any>()
        map["toUserID"] = userInfoModel.userId
        map["worksID"] = model.worksID
        val body = RequestBody.create(MediaType.parse(ApiManager.APPLICATION_JSON), JSON.toJSONString(map))
        ApiMethods.subscribe(mUserInfoServerApi.playWorks(body), object : ApiObserver<ApiResult>() {
            override fun process(result: ApiResult) {
                if (result.errno == 0) {
                    // TODO: 2019/5/22 播放次数客户端自己加一
                    model.playCnt = model.playCnt + 1
                    mAdapter.notifyDataSetChanged()
                }
            }
        }, mFragment, RequestControl("playWorks", ControlType.CancelThis))
    }

    private fun addProducation(list: List<ProducationModel>?, newOffset: Int, totalCnt: Int, isClear: Boolean) {
        offset = newOffset
        mLastUpdateInfo = System.currentTimeMillis()

        hasMore = !list.isNullOrEmpty()
        mCallBack?.onRequestSucess(hasMore)

        if (isClear) {
            mAdapter.dataList.clear()
        }

        if (list != null && list.isNotEmpty()) {
            mAdapter.dataList.addAll(list)
            mAdapter.notifyDataSetChanged()
        } else {
            if (mAdapter.dataList != null && mAdapter.dataList.size > 0) {
                // 没有更多了
            } else {
                // 没有数据
            }
        }
    }

    private fun loadProducationsFailed() {
        mCallBack?.onRequestSucess(true)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        SinglePlayer.stop(playerTag)
        if (mShareWorksDialog != null) {
            mShareWorksDialog!!.dismiss(false)
        }
        if (mConfirmDialog != null) {
            mConfirmDialog!!.dismiss(false)
        }
    }

    fun destory() {
        SinglePlayer.reset(playerTag)
        SinglePlayer.removeCallback(playerTag)
    }
}
